### ISSUE TITLE
Support java.util.Map as a mappable object

### ISSUE THREAD

**Issue by igarashitm** [atlasmap/atlasmap#142]
The option in my mind is to
* Add `FieldType.DYNAMIC_MAP` and link with `java.util.Map`. So as a Java Document field, it's just a field, won't look into its contents
* Only for a `atlas:core:DYNAMIC_MAP` DataSource (#9), provide a way to feed keys/types list in UI to prefill the field list, but also allows to add new field manually
  * As a 1st path. We might consider doing it even as a part of Java Document.

**Comment by igarashitm** [comment]
@mattrpav what do you think about this?

**Comment by igarashitm** [comment]
A bit background - we want a bucket for carrying headers/properties like things into atlasmap, and it's the initial usecase for the multiple source/target support (#64). Problem is that those are all dynamic `java.util.Map`, sometimes some of keys can be explicitly prefilled at design time, but not all.

**Comment by mattrpav** [comment]
@igarashitm I'd recommend looking into accomplishing it without a different FieldType. I think it would just be root object is java.util.Map .. then provide an actual concrete class for Target side.

AtlasMap coordinates should be able to address the uses cases:

"{}" was intended to specify map keys

/{foo}
/{bar}
/{blah}
/{} <-- all keys
.. etc

**Comment by mattrpav** [comment]
Possible use cases:

1. Single source named Map field referenced by key to a single target field:   /{foo} -> /targetObject/targetField
2. All Source Map fields referenced by key to target field:   /{} -> /targetObject/targetList<> 
    aka. Map keys to a List<String>
3. Single source field to a single named Map key field /sourceObject/sourceField -> /{foo}
4. Multiple source fields to Map target keys /sourceObject/sourceList<> -> /{}

Others?

**Comment by igarashitm** [comment]
Ah sounds good. Do you mean by No.2 and No.4, just pack keys by the `{}` suffix? Do we want to another suffix to mean "all values"?

**Comment by igarashitm** [comment]
And what if we want to just carry whole map? like: /{} -> /targetObject/parameters{}

**Comment by mattrpav** [comment]
Yes, just like arrays and lists work today.. /foo[] and /bar<>..  "/{}" or "/bar{} would mean all map values

**Comment by mattrpav** [comment]
Relevant IRC discussion:

> igarashitm: mattrpav: and I took over the 9 mainly because to introduce DYNAMIC_MAP field type, but if we don't go that path I would take it back to you
[11:10am] mattrpav: did you see my note about DYNAMIC_MAP?
[11:10am] mattrpav: i’m not sure you need a special type
[11:10am] igarashitm: yep, and made some more comments there, generally you idea looks good
[11:11am] igarashitm: a bit more work to implement though
[11:11am] igarashitm: compared to just introducing DYNAMIC_MAP field type
[11:12am] mattrpav: i think the collection hanlding already supports looping for lists and arrays
[11:12am] mattrpav: i think the big question is— is a MAP a collection or a field type?
[11:13am] mattrpav: i think its a collection.. b/c inside each value can be different field types
[11:13am] mattrpav: then FieldActions and converters that support Collection.MAP or Collection.ALL can implement support for Map, Array or List as they see fit
[11:14am] igarashitm: ah, good point
[11:14am] igarashitm: I'm wondering how it looks in UI
[11:15am] igarashitm: with that path, is it still possible to expand the Map object and show field list by providing a pre-fill list for the Map?
[11:15am] igarashitm: list of keys/types
[11:15am] mattrpav: https://github.com/atlasmap/atlasmap/blob/master/runtime/core/src/main/java/io/atlasmap/core/BaseAtlasModule.java#L115
[11:17am] mattrpav: ah gotcha.. yeah.. so for UI display i think we can re-use a LookupTable or PropertyField type UX behavior.. if I’m following you.. you’d want ot inject a list of predefined KEYS, but not values to the UI, correct?
[11:17am] igarashitm: yep that's it, and then also allow to add a user defined entry by putting a key
[11:18am] mattrpav: i think a list of map keys is pretty similar to how we’d do enum.. so supporting a design-time only lookup table or similar for reference would do the trick
[11:18am] mattrpav: lookuptable 
[11:18am] mattrpav: all day

**Comment by mattrpav** [comment]
Use cases to consider:

1. UI needs to inject a set of 'known' keys in a map
2. UI needs to allow user to specify their own 'key' for use in a map

Suggested approach: look at using a LookupTable and having the syndesis UI inject a LookupTable to the mapping at design time.

**Comment by igarashitm** [comment]
Hi @mattrpav ,

So it's a volatile `MapLookupTable` model to provide a UI form to handle `Map` keys mapping in UI, but not persist into mapping file. Is it correct? When it persists the mapping file, fixed `Map` keys  with like `{foo}` (or whole `Map` by `{}`) are there. As actual `LookupTable in the mapping file is already used for different purpose.

Then, how `FieldType` should look like for the non-primitive values? If it's `COMPLEX`, inspection/FieldReader/FieldWriter would still drill down into its field I guess?

**Comment by igarashitm** [comment]
I think in the first implementation it would be OK to only allow atlas-primitives for the `Map` values. We may want to got that path anyway, and consider `COMPLEX` handling in general. It's gonna be a huge topic, involving `TypeConverter`/`FieldAction` for the `COMPLEX` type I guess.

**Comment by mattrpav** [comment]
Agreed.. I think you are clarifying use cases. In general, I think the approach of Map being a 'collection', just like Array and List is a good guiding principal.  Just like Array and List, we started with primitives being the only supported member types, and then worked our way up to more complex scenarios.

I think we should define very specific use cases, and then build the code and tests according to product release priorities and real-world use cases that pop-up. This has the side benefit to  keep change sets small as well.

Note: All Map use cases tasks should require design, testing and validating:

1. AtlasPath syntax (and potentially enhancement)
2. Potential changes to the AtlasMapping.Collection object (perhaps adding the ability to reference a LookupTable)
2. Field Actions that support Collection.MAP or Collection.ANY
3. TypeConverters that support Collection.MAP or Collection.ANY

We could create separate git issues for each of:

Map UC-1: Map<String, String> .. 
Map UC-2: Map<String, .. any Atlas.PRIMITIVE> ..
Map UC-3: Map<.. any Atlas.PRIMITIVE, .. any Atlas.PRIMITIVE> ..
Map UC-4: Map<String, Atlas.COMPLEX>..
Map UC-5: Map< .. any Atlas.Primitive.. , Atlas.COMPLEX>..
Map UC-6: Map <Atlas.COMPLEX, Atlas.COMPLEX> ..

**Comment by mattrpav** [comment]
Re the UI: I understand that the UI builds a LookupTable for enum support and saves it to the AtlasMapping definition file. The UI has a volatile model, but the UI should also be saving the final saved data into the AtlasMapping definition.

**Comment by igarashitm** [comment]
Yeah, at first `<String, String>` should be fine as it's path parameter for the REST, then at some point we'll need <String, Object> which is coming from Camel Message headers.
